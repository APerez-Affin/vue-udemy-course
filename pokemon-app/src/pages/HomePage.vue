<template>
  <div v-if="!pokemon">
    <h1>Espere por favor</h1>
  </div>
  <div v-else>
    <h1>¿Quién es este pokemón?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selected-pokemon="checkAnswer" />

    <div v-if="showAnswer">
      <h2 class="fadeIn">{{ message }}</h2>
      <button @click="newGame">Volver a intentar</button>
    </div>
  </div>
</template>

<script>
  import { PokemonOptions, PokemonPicture } from "@/components";
  import getPokemonsOptions from "@/helpers/getPokemonOptions";
  getPokemonsOptions();
  export default {
    components: { PokemonOptions, PokemonPicture },
    name: "HomePage",
    data() {
      return {
        pokemonArr: [],
        pokemon: null,
        showPokemon: false,
        showAnswer: false,
        message: "",
      };
    },
    methods: {
      async mixPokemonArray() {
        this.pokemonArr = await getPokemonsOptions();
        const randInt = Math.floor(Math.random() * 4);
        this.pokemon = this.pokemonArr[randInt];
      },
      checkAnswer(id) {
        this.showPokemon = true;
        this.showAnswer = true;
        id === this.pokemon.id
          ? (this.message = `Correcto, es ${this.pokemon.name}`)
          : (this.message = `Oops, era ${this.pokemon.name}`);
      },
      newGame(){
        this.showPokemon = false;
        this.showAnswer = false;
        this.message = '';
        this.pokemon= null;
        this.mixPokemonArray();
      }
    },
    mounted() {
      this.mixPokemonArray();
    },
  };
</script>

<style></style>
