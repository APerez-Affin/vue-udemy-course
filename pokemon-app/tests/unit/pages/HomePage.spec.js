import { shallowMount, mount } from "@vue/test-utils";
import HomePage from "@/pages/HomePage";

import { pokemons } from "../mocks/pokemons.mock";

describe("HomePage Component", () => {
  let wrapper;
  let mixPokemonArraySpy;
  // Carga el componente antes de ejecutar los test con el mock de pokemons
  beforeEach(() => {
    mixPokemonArraySpy = jest.spyOn(HomePage.methods, "mixPokemonArray");
    wrapper = shallowMount(HomePage);
  });

  test("Should match with the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Should call mixPokemonArray", () => {
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("Should match with the snapshot when the pokemons are loaded", () => {
    const wrapper = shallowMount(HomePage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Should show the components PokemonPicture && PokemonOptions", () => {
    const wrapper = shallowMount(HomePage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    const picture = wrapper.find("pokemon-picture-stub");
    const options = wrapper.find("pokemon-options-stub");

    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    expect(picture.attributes("pokemonid")).toBe("1");
    expect(options.attributes("pokemons")).toBeTruthy();
  });

  test("Test with checkAnswer", async () => {
    const wrapper = shallowMount(HomePage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    await wrapper.vm.checkAnswer(10);

    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`);
  });
});
