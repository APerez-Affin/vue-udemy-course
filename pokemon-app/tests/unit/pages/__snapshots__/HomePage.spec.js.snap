// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component Should match with the snapshot 1`] = `
<div>
  <h1>Espere por favor</h1>
</div>
`;

exports[`HomePage Component Should match with the snapshot when the pokemons are loaded 1`] = `
<div>
  <h1>¿Quién es este pokemón?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
